plugins/libplugin-pay.c

```diff
@@ -575,8 +575,12 @@ static struct command_result *payment_finished(struct payment *p)
 	assert((result.leafstates & PAYMENT_STEP_SUCCESS) == 0 ||
 	       result.preimage != NULL);
 
-	if (p->parent == NULL) {
-		assert(p->cmd != NULL);
+	if (p->parent == NULL && cmd == NULL) {
+		/* This is the tree root, but we already reported success or
+		 * failure, so noop. */
+		return command_still_pending(NULL);
```

Comment:
Yech, that's ugly!  I'm not convinced payment_finished should retrurn command_result...

Reply:
The reason it is like this is that `payment_finished` encapsulates the logic to determine whether a command is pending or not. It is the place where we call `command_fail` and `command_finished`. Those two functions are marked as `warn_unused_result`, so I thought it easiest to hand them up and then silence the result.

We can of course silence it by doing something else, but I have no good ideas on how to do that.

Reply:
After some more googling it seems like we can opt-out of the `warn_unused_result` using a cast to void (???):

```c
(void)command_finished(cmd, ret);
```

Still not nice, but I added what they'd look like in the last few `fixup!`-commits (they clash with other things so I haven't reordered them just yet).

Reply:
Hm, `gcc` (7.5) doesn't seem to like this trick, and still warns about the unused result.

Reply:
No, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425 :(